EIPS/eip-4511.md

```diff
@@ -0,0 +1,116 @@
+
+Because of this, BEAM bytecode is highly amenable to optimization techniques such as threaded-code interpretation and Just-In-Time compilation. The codegen for individual bytecode instructions can often be as simple as two host-ISA instructions: one to do the required math/stack manip/etc., and the other to subtract an appropriate amount from the reduction counter.
+
+Even better, after codegen, peephole optimization can be applied to hoist all these reduction-counter decrements (which only depend on each-other) out of the basic block they're embedded in, and replace them with a single reduction-counter decrement of the aggregate value from the individual operations, drastically reducing generated code size, and enabling other optimizations due to the code now forming clear "runs" of dependent instructions, without any interruptions from "bookkeeping" code.
+
+### Considerations in Borrowing the BEAM Approach
+
+Note that branching instructions are not included in the set of BEAM instructions that check the reduction counter. This is because the Erlang (and other BEAM language) compilers are careful to never generate branches that represent *back edges* in a Control Flow Graph. Because of this, despite primitive looping constructs being theoretically possible in BEAM bytecode, they're never emitted in practice.
+
+Instead, Erlang and other BEAM languages expose Tail-Call-Optimized (TCOed) recursion as a primitive, and looping constructs are implemented in terms of that primitive. As recursive tail-calls use the "subroutine call" BEAM instruction, these calls always flow through the Out-of-Reductions check-and-branch logic built into that instruction's implementation.
+
+Because of this, BEAM VM languages are free to use *forward* branch instructions to implement non-loop control-structures, e.g. `if/else`, `case`, clause-head pattern matching, `goto fail`-type internal error handling, etc., without paying any bookkeeping overhead cost for these instructions. The instruction implementations for forward-branches remain just as simple as the instruction implementations for math ops *et al*, and so become amenable to optimization (e.g. branchless-code reduction) themselves.
+
+The EVM, as a trustless-multitenant runtime, cannot rely on the developer or compiler kindly limiting themselves to only deploying code that uses a subset of possible instruction sequences, the way BEAM compilers do. The EVM must ensure that Out-of-Gas conditions are checked at least once for each iteration of EVM-primitive branching loops.
+
+## Specification
+
+### Definitions
+The key words “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”, “SHOULD NOT”, “RECOMMENDED”, “MAY”, and “OPTIONAL” in this document are to be interpreted as described in RFC 2119.
```

Comment:
```suggestion
```
This is implied by EIP-1.

Reply:
We recently added this to the template, and I believe the consensus was that it should be included in every EIP?  @axic may remember better.  If we *do* want to say it is implied by EIP-1, then we should remove it from the template.

Reply:
Yep that's my mistake. Was thinking the RFC 2119 section in EIP-1 implied all EIPs were under that. We can leave as-is and I'll revert these requests.